src/nvme/mi.h

```diff
@@ -955,6 +955,44 @@ int nvme_mi_admin_xfer(nvme_mi_ctrl_t ctrl,
 		       off_t resp_data_offset,
 		       size_t *resp_data_size);
 
+/**
+ * nvme_mi_admin_admin_passthru() - Submit an nvme admin passthrough command
+ * @ctrl: Controller to send command to
+ * @opcode:	The nvme admin command to send
+ * @flags:	NVMe command flags (not used)
+ * @rsvd:	Reserved for future use
+ * @nsid:	Namespace identifier
+ * @cdw2:	Command dword 2
+ * @cdw3:	Command dword 3
+ * @cdw10:	Command dword 10
+ * @cdw11:	Command dword 11
+ * @cdw12:	Command dword 12
+ * @cdw13:	Command dword 13
+ * @cdw14:	Command dword 14
+ * @cdw15:	Command dword 15
+ * @data_len:	Length of the data transferred in this command in bytes
+ * @data:	Pointer to user address of the data buffer
+ * @metadata_len:Length of metadata transferred in this command(not used)
+ * @metadata:	Pointer to user address of the metadata buffer(not used)
+ * @timeout_ms:	How long to wait for the command to complete
+ * @result:	Optional field to return the result from the CQE dword 0
+ *
+ * Send a customized NVMe Admin command request message and get the corresponding
+ * response message.
+ *
+ * This command only supports data transfer size range [0, 4096] (bytes) so
+ * the & data_len paramter must be less than 4097.
+ *
+ * Return: The nvme command status if a response was received (see
+ * &enum nvme_status_field) or -1 with errno set otherwise.
+ */
+int nvme_mi_admin_admin_passthru(nvme_mi_ctrl_t ctrl, __u8 opcode, __u8 flags,
+				 __u16 rsvd, __u32 nsid, __u32 cdw2, __u32 cdw3,
+				 __u32 cdw10, __u32 cdw11, __u32 cdw12,
+				 __u32 cdw13, __u32 cdw14, __u32 cdw15,
+				 __u32 data_len, void *data, __u32 metadata_len,
+				 void *metadata, __u32 timeout_ms, __u32 *result);
+
```

Comment:
`struct nvme_passthru_cmd`  ?

Reply:
The in-band API `nvme_admin_passthru` and `nvme_admin_passthru64` are still using the same long argument list.  If we  use `struct nvme_passthru_cmd` here, we need to change the in-band APIs too. I'm not sure whether in-band API change will cause any dependency issues.

Reply:
Ah, okay, I thought we were using the `struct nvme_passthru_cmd` for those commands as well. I suppose we just keep it consistent for this interface.